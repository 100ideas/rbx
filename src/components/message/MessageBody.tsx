import { cx } from "emotion";
import React from "react";

import Element from "components/element";
import renderAsExoticComponent from "components/render-as-exotic-component";
import { ModifierProps } from "modifiers";

export type MessageBoodyModifierProps = Partial<{
  children: React.ReactNode;
}>;

export type MessageBodyProps = ModifierProps & MessageBoodyModifierProps;

const MessageBody = renderAsExoticComponent<MessageBodyProps, "div">(
  ({ children, className, ...props }, ref) => (
    <Element {...props} ref={ref} className={cx("message-body", className)}>
      {children}
    </Element>
  ),
  "div",
);
MessageBody.defaultProps = Object.assign(
  {
    children: null,
  },
  MessageBody.defaultProps,
);

export default MessageBody;
