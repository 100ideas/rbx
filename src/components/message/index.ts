import "./message.module.sass";

export { Message } from "./message";
