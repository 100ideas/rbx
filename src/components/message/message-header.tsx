import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/generic";
import { ModifierProps, transformModifiers } from "@/modifiers";

export type MessageHeaderProps = ModifierProps;

export const MessageHeader = forwardRefAs<MessageHeaderProps, "div">(
  (props, ref) => {
    const { as, ...rest } = transformModifiers(props);
    rest.className = cx("message-header", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);
