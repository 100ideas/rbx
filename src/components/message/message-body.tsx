import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type MessageBodyProps = ModifierProps;

export const MessageBody = asExoticComponent<MessageBodyProps, "div">(
  (props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("message-body", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);
