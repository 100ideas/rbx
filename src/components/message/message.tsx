import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";
import { Colors } from "@/modifiers/color";
import { MessageBody } from "./message-body";
import { MessageHeader } from "./message-header";

export type MessageModifierProps = Partial<{
  color: Colors;
  size: "small" | "medium" | "large";
}>;

export type MessageProps = ModifierProps & MessageModifierProps;

export const Message = Object.assign(
  asExoticComponent<MessageProps, "article">((props, ref) => {
    const { as, color, size, ...rest } = modify(props);
    rest.className = cx("message", rest.className, {
      [`is-${color}`]: color,
      [`is-${size}`]: size,
    });
    return React.createElement(as!, { ref, ...rest });
  }, "article"),
  { Body: MessageBody, Header: MessageHeader },
);
