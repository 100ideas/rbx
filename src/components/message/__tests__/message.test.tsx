import React from "react";
import renderer from "react-test-renderer";

import { Message } from "../message";

describe("Message component", () => {
  it("should Exist", () => {
    expect(Message).toMatchSnapshot();
  });

  it("should have message classnames", () => {
    const component = renderer.create(
      <Message>
        <Message.Header>Lorem Ipsum</Message.Header>
        <Message.Body>Lorem Ipsum</Message.Body>
      </Message>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat Bulma class with classes in props", () => {
    const component = renderer.create(
      <Message className="other-class">
        <Message.Header>Lorem Ipsum</Message.Header>
        <Message.Body>Lorem Ipsum</Message.Body>
      </Message>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render as an html section", () => {
    const component = renderer.create(
      <Message<"section"> as="section">This should be a section</Message>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should have custom inline styles", () => {
    const component = renderer.create(
      <Message<"section"> as="section" style={{ width: 200, zIndex: 1 }}>
        This should be a section with custom styles
      </Message>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should accept a react Element as renderAs prop", () => {
    const Custom = (props: React.HTMLAttributes<HTMLParagraphElement>) => (
      <p {...props}>
        Custom
        {props.children || null}
      </p>
    );

    const component = renderer.create(
      <Message<typeof Custom> as={Custom}>This should be a p element</Message>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
