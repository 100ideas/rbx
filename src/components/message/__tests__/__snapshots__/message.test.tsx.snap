// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component should Exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Body": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "Header": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "defaultProps": Object {
    "as": "article",
  },
  "render": [Function],
}
`;

exports[`Message component should accept a react Element as renderAs prop 1`] = `
<p
  className="message"
>
  Custom
  This should be a p element
</p>
`;

exports[`Message component should concat Bulma class with classes in props 1`] = `
<article
  className="message other-class"
>
  <div
    className="message-header"
  >
    Lorem Ipsum
  </div>
  <div
    className="message-body"
  >
    Lorem Ipsum
  </div>
</article>
`;

exports[`Message component should have custom inline styles 1`] = `
<section
  className="message"
  style={
    Object {
      "width": 200,
      "zIndex": 1,
    }
  }
>
  This should be a section with custom styles
</section>
`;

exports[`Message component should have message classnames 1`] = `
<article
  className="message"
>
  <div
    className="message-header"
  >
    Lorem Ipsum
  </div>
  <div
    className="message-body"
  >
    Lorem Ipsum
  </div>
</article>
`;

exports[`Message component should render as an html section 1`] = `
<section
  className="message"
>
  This should be a section
</section>
`;
